inter/pos/validators.go

```diff
@@ -29,108 +36,104 @@ type (
 	GValidators map[idx.StakerID]GenesisValidator
 )
 
+func NewBuilder() ValidatorsBuilder {
+	return ValidatorsBuilder{}
+}
+
+// Set appends item to Validator object
+func (vv ValidatorsBuilder) Set(id idx.StakerID, stake Stake) {
+	// add
+	if stake == 0 {
+		delete(vv, id)
+	} else {
+		vv[id] = stake
+	}
+}
+
+func (vv ValidatorsBuilder) Build() *Validators {
+	return NewValidators(vv)
+}
+
 // NewValidators return new pointer of Validators object
-func NewValidators() *Validators {
-	return &Validators{
-		indexes: make(map[idx.StakerID]int),
-		list:    make([]Stake, 0, 200),
-		ids:     make([]idx.StakerID, 0, 200),
+func NewValidators(values ValidatorsBuilder) *Validators {
+	valuesCopy := make(ValidatorsBuilder)
+	for id, s := range values {
+		valuesCopy.Set(id, s)
 	}
+
+	vv := &Validators{
+		values: valuesCopy,
+	}
+	vv.cache = vv.calcCaches()
+	return vv
 }
 
 // Len return count of validators in Validators objects
-func (vv Validators) Len() int {
-	return len(vv.list)
+func (vv *Validators) Len() int {
+	return len(vv.values)
 }
 
-// Iterate return chanel of ids for get validators in loop
-func (vv Validators) Iterate() <-chan idx.StakerID {
-	c := make(chan idx.StakerID)
-	go func() {
-		for _, a := range vv.ids {
-			c <- a
-		}
-		close(c)
-	}()
-	return c
+// Iterate return slice of ids for get validators in loop
+func (vv *Validators) Iterate() []idx.StakerID {
+	return vv.cache.ids
 }
 
-// Set appends item to Validator object
-func (vv *Validators) Set(id idx.StakerID, stake Stake) {
-	if stake != 0 {
-		i, ok := vv.indexes[id]
-		if ok {
-			vv.list[i] = stake
-			return
-		}
-		vv.list = append(vv.list, stake)
-		vv.ids = append(vv.ids, id)
-		vv.indexes[id] = len(vv.list) - 1
-	} else {
-		i, ok := vv.indexes[id]
-		if ok {
-			delete(vv.indexes, id)
-			idxOrig := len(vv.list) - 1
-			if i == idxOrig {
-				vv.list = vv.list[:idxOrig]
-				vv.ids = vv.ids[:idxOrig]
-			} else {
-				// Move last to deleted position + truncate list len
-				vv.list[i] = vv.list[idxOrig]
-				vv.list = vv.list[:idxOrig]
-
-				vv.indexes[vv.ids[idxOrig]] = i
-
-				vv.ids[i] = vv.ids[idxOrig]
-				vv.ids = vv.ids[:idxOrig]
-			}
-		}
+func (vv *Validators) calcCaches() cache {
+	cache := cache{
```

Comment:
do we need to recache it somewhere? It seems it's always initialized once.
Change 'cache' to a better name?